core/src/main/java/org/apache/calcite/rel/rules/PushProjector.java

```diff
@@ -216,9 +218,14 @@ public PushProjector(
       origProjExprs = origProj.getProjects();
     }
 
-    childFields = childRel.getRowType().getFieldList();
+    if (childRel instanceof Join) {
+      Join join = (Join) childRel;
+      childFields = Lists.newArrayList(join.getLeft().getRowType().getFieldList());
+      childFields.addAll(join.getRight().getRowType().getFieldList());
+    } else {
+      childFields = childRel.getRowType().getFieldList();
+    }
     nChildFields = childFields.size();
```

Comment:
What is the difference if childRel is join?

Reply:
When childRel is a join, the row type of childRel may not be sufficient. 
This can only happen when the join type is semi or anti.

For such cases, the row type of childRel only contains the fields of the left input. However, we need fields from both inputs. This is because, the join conditions may reference columns from the right, even if the row type does not involve the right side.

Reply:
Thanks for your explanation.